### DISCUSSION TITLE
How to generate cask json without homebrew installed?

### DISCUSSION THREAD

**Discussion by ldeck** [Homebrew/discussions#2861] [Casks]
As I understand it from #2803 , the homebrew JSON API is generated by evaluating a cask and producing the equivalent json.

The problem is that as this is currently done on an intel machine, it's not currently possible to obtain the json for when `Hardware::CPU.intel?` is false.

I'm not familiar enough with ruby, but wondered if it's possible to do something like the following to obtain the json for the false case? Can this be done without homebrew installed? e.g., could the cask be evaluated or translated to something else and then evaluated in plain ruby that could be evaluated to json?

```ruby
#!/usr/bin/env ruby

require 'open-uri'

cask = ARGV[0]
cask_uri = "https://raw.githubusercontent.com/Homebrew/homebrew-cask/master/Casks/#{cask}.rb"

cask_content = URI.open(cask_uri, &:read).to_s
content = cask_content
  .gsub(/Hardware::CPU.intel\?/, 'false')
  .to_s

json = # somehow eval the cask to json?
puts json
```

What I'm trying to do is translate the ruby based cask to another representation. In my case to a nix home-manager module. But others may have other use-cases.

Thanks!

**Comment by carlocab** [comment]
It's possible, but you're probably going to end up re-writing significant parts of Homebrew (at least the `cask` parts of it) in the process. But maybe not too much of it if all you're interested in is metadata.

Note that the strategy you've outlined likely opens you up to remote-code execution though. That's probably not what you want.

It might be simpler to fix `brew` to generate the metadata that you want. See Homebrew/brew#12786.

**Reply by ldeck** [reply]
@carlocab — yes, to clarify, I'm only interested in the metadata. 
Nixpkgs will handle the rest. See [ldeck/nix-home/.../lib/defaults/apps/macOS/chromium.nix](https://github.com/ldeck/nix-home/blob/05ea5db17f5f1d8dcd81c0f257eaa5fca898bd65/lib/defaults/apps/macOS/chromium.nix) for example.

I'd not seen the issue you've linked to, but had raised a similar issue as mentioned about the problem with the json api not making the hardware variants metadata available. That will take some fixing, given the number of casks checking specifically for the hardware they're running on without any abstraction.

So, given this was a known issue, I was looking to move ahead quickly in the meantime rather than wait for the json api to be fixed. No doubt when it's fixed it'll be highly useful.

You're right, however, that remote-code execution could be a problem. I'd overlooked that. So that leads me back to parsing the ruby code instead I suppose which I was initially hoping to avoid.

**Reply by carlocab** [reply]
Might not be completely hopeless -- all of the arch-specific conditionals are (intentionally) of the form
```ruby
if Hardware::CPU.intel?
  ...
else
  ...
```
so parsing them might turn out to be not too hard. I don't think (or know if) we have CI checks for this style, though (we have them for most other style conventions), so you might want to double-check this.
